alchemy/src/vercel/project-domain.ts

```diff
@@ -0,0 +1,144 @@
+import type { Context } from "../context.js";
+import { Resource } from "../resource.js";
+import type { Secret } from "../secret.js";
+import { createVercelApi } from "./api.js";
+
+/**
+ * Properties for creating or updating a ProjectDomain
+ */
+export interface ProjectDomainProps {
+  /**
+   * The domain name to add to the project
+   */
+  name: string;
+
+  /**
+   * The unique project identifier or the project name
+   */
+  project: string;
+
+  /**
+   * The Git branch to link the domain to
+   */
+  gitBranch?: string;
+
+  /**
+   * The unique custom environment identifier within the project
+   */
+  customEnvironment?: string;
+
+  /**
+   * The redirect target for the domain
+   */
+  redirect?: string;
+
+  /**
+   * The redirect status code
+   */
+  redirectStatusCode?: 301 | 302 | 307 | 308;
+}
+
+/**
+ * Output returned after ProjectDomain creation/update
+ */
+export interface ProjectDomain
+  extends Resource<"vercel::ProjectDomain">,
+    ProjectDomainProps {
+  apexName: string;
+
+  /**
+   * `true` if the domain is verified for use with the project. If `false` it will not be used as an alias on this project until the challenge in `verification` is completed.
+   */
+  verified: boolean;
+
+  /**
+   * The time at which the domain was created
+   */
+  createdAt: number;
+
+  /**
+   * The time at which the domain was last updated
+   */
+  updatedAt: number;
+
+  /**
+   * A list of verification challenges, one of which must be completed to verify the domain for use on the project. After the challenge is complete `POST /projects/:idOrName/domains/:domain/verify` to verify the domain.
+   *
+   * Possible challenges:
+   *   - If `verification.type = TXT` the `verification.domain` will be checked for a TXT record matching `verification.value`.
+   */
+  verification?: {
+    /**
+     * The type of verification
+     */
+    type: string;
+
+    /**
+     * The domain being verified
+     */
+    domain: string;
+
+    /**
+     * The verification value
+     */
+    value: string;
+
+    /**
+     * The reason for verification
+     */
+    reason: string;
+  }[];
+}
+
+export const ProjectDomain = Resource(
+  "vercel::ProjectDomain",
+  async function (
+    this: Context<ProjectDomain>,
+    id: string,
+    { accessToken, ...props }: ProjectDomainProps & { accessToken?: Secret },
+  ): Promise<ProjectDomain> {
+    switch (this.phase) {
+      case "create": {
+        const api = await createVercelApi({
+          baseUrl: "https://api.vercel.com/v10",
+          accessToken,
+        });
+        const domain = (await api
+          .post(`/projects/${props.project}/domains`, props)
+          .then((res) => res.json())) as ProjectDomain;
+
+        return this.create({ ...props, ...domain });
+      }
+
+      case "update": {
```

Comment:
I usually don't like switch but it works well here.

Reply:
I go back & forth.  My OOP preference is explicit `create()` methods that have to be defined, but these large blocks can be refactored any number of ways 💪